### PR TITLE
Typescript Support

### LINKED ISSUE #542: Typescript
As I am trying to read the tokenize/tsc parts and trying to do some changes/experiments. I found myself struggling on finding the token names and what's going on. 

I would propose we switch to Typescript for better DX. XD
Yes, I know this is huge. But please kindly consider this.

Benefits
- Strong types, fewer bugs.
- Better editor support, make easier for contributors to help
- Exported typing files for packages. ( compiles or other stuff get a lot of options, with typing file, editors can do auto-complete and inline documentation, and even for js users)
- Use newer js features and transpile to the current env.
- For the benefits of Typescript, I can talk for a whole day. You can also check out https://medium.com/@jtomaszewski/why-typescript-is-the-best-way-to-write-front-end-in-2019-feb855f9b164

Cost
- Toolings. Since we are already using webpack, it's fairly simple to integrate. I can handle that.
- Some code refactoring.
- A little learning but not steep

I am doing to experiment on a branch. Let me know what you think before I continue. Thanks!

### PR DESCRIPTION
**Pull request by antfu**
close #542 

Ok ok. I just can't control my hand, so here I am. 🤪 Luckily there is not too much work to do for the migration. Please don't feel pressure about this. Try it out when you got some time and make the decision when you feel good. 

## Changes
- Use Typescript!
- Convert most of js code to ts
- Switch to Jest from Mocha to better work with ts
- All tests passed

## Todos
- [x] Types for tokens and ASC
- [x] `.d.ts` typing files build

## Notes
- I won't rewrite `highlight.js` to ts since it only works for legacy ide and should be removed eventually. 
- I move pinyin and dexter mapping table to json files. In order to make code cleaner.
- Renamed the compiler folder to `transpiler` since it's kinda ambiguous with `compile` in `parser`
- I actually found some edge bugs during this migration, and I fixed them. (thx ts XD)

### CODE REVIEW

**Comment by LingDong-** [comment]
👍Wow that's quick :D
I took a peek and it seems promising. Some quick questions about the current workflow:

- Just checking, it seems that to build the thing I just go `npm run build` and typescript, webpack and all the other stuff just goes automatically kicks in and I'll get just the minified output in `./dist`?
- Sometimes I just want to run 1 file and check some quick outputs. Before it was `node hello.js`. For typescript I figure I can do this:

run.sh:
```bash
tsc $1.ts; node $1.js; rm ./*.js;
```
and then:
```
run hello
```
Not sure if it's the best way :P I bet VSCode have some functionality to do this already. And I think I need to get `tsc` to output some source-map which I then need to get IDE to understand so they can spit meaningful error messages... Lot's of stuff to learn!

Speaking of source maps, I'm thinking maybe wenyan-lang can use some source mapping too, do you think? I can look into it.

So far I think this is looking good. But let me research more into TypeScript first.

<sub>
Some rants: One of the reasons I like statically typed languages (besides compile-time errors) is that they run fast. So while I'm tirelessly typing all those types I feel my labor is useful. But then for TypeScript it just strips out my type annotations as it compiles. I feel like my work is getting thrown away! XD And it just compiles back to JavaScript and runs just about as slow! It's also like learning a new thing, and all the new thing does is to convert itself back to the old thing I already knew. 
</sub>
<br>
<sub>
The compile time type checking is definitely very helpful for debugging, but as I've worked with these duck-typed languages a lot I sort of taught my brain to do what the compiler does. I guess the biggest benefit for me would be to have self-documenting code, and of course, make collaboration easier.
</sub>
<br>
<sub>
I'm not saying TypeScript is not great, I'm just hoping Web is TypeScript instead of JavaScript all along. It's just too late. It's like first having a cat, but realizing what people actually need is a tiger, go through all those plastic surgery to make the cat look like a tiger (which then compiles itself back to the cat). I really hope we get a tiger in the first place.
</sub>
<br>
<sub>
But anyways, rant aside, I still think TypeScript is worth a try. If we convert, let's add type annotation to every single thing so that we make best use of what TypeScript offers, then it'll be worth the trouble and overhead in conversion.
</sub>

**Comment by antfu** [comment]
> Just checking, it seems that to build the thing I just go npm run build and typescript, webpack and all the other stuff just automatically kick in and I'll get just the minified output in ./dist?

Basically, yes. But typescripts also has a thing called [Decalarion File](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) which allows compiled js to keep their typing for editors while making he code functional for node and browsers. You can check out the `dist/code/index.d.ts` after `npm run build` for example.

> Sometimes I just want to run 1 file and check some quick outputs. Before it was node hello.js. For typescript I figure I can do this

Try [`ts-node`](https://github.com/TypeStrong/ts-node)! It's awesome. `ts-node hello.ts`!

> And I think I need to get tsc to output some source-map 

Sure. That's enabled that! I turned it off in the first place in order to make the package relatively small XP

> Speaking of source maps, I'm thinking maybe wenyan-lang can use some source mapping too, do you think? I can look into it.

That will be cool. Actually I don't know much about source maps. For me, it's just a magic that works XD. Let's try it. Good time to learn :)

>  I'm just hoping Web is TypeScript instead of JavaScript all along. It's just too late.

That's what all of us expected to see in the future!

>  I really hope we get a tiger in the first place.

Yes. Of course. But I would say I am very grateful for what it is now. I saw that js from a "toy" language become something that has great power and runs everywhere. I haven't seemed anything grows so fast as the web did. And it should be credited to the great engineering of open source community. I think the cat will eventually become the tiger or something bigger. Looking forward to that! 😉

File: test/utils.ts [hunk 1]
```diff
@@ -46,15 +52,20 @@ function compileLib(name, options = {}) {
...
+      expect(actual === expected).toBeTruthy();
+    }
+  } catch (e) {
+    if (message) throw new Error(message);
+    else throw e;
   }
 }
 
-function compareResults(x, y) {
+function compareResults(x: number, y: number) {
```
Comment (statementreply) [inline test/utils.ts]: `x: number | number[], y: number | number[]`

Sorry for the lack of comments :|

File: test/utils.ts [hunk 2]
```diff
@@ -72,30 +83,43 @@ function compareResults(x, y) {
   return 0;
 }
 
-function isOnCorrectSide(actual, expected, boundary) {
-  return compareResults(actual, boundary) * compareResults(expected, boundary) >= 0;
+function isOnCorrectSide(actual: number, expected: number, boundary: number) {
```
Comment (statementreply) [inline test/utils.ts]: `actual: number | number[], expected: number | number[], boundary: number | number[]`

File: test/utils.ts [hunk 3]
```diff
@@ -72,30 +83,43 @@ function compareResults(x, y) {
...
     return expected.reduce((err, y) => err - y, actual);
   } else {
     return actual - expected;
   }
 }
 
-function nearlyEqual(actual, expected, options = {}) {
+function nearlyEqual(
+  actual: number,
+  expected: number,
```
Comment (statementreply) [inline test/utils.ts]: `expected: number | number[]`

File: test/utils.ts [hunk 4]
```diff
@@ -118,21 +142,34 @@ function nearlyEqual(actual, expected, options = {}) {
  *     The actual result shall not be on the different side of the boundary then the expected.
  * @param {string} [message] - Debug message.
  */
-function assertNearlyEqual(actual, expected, options = {}, message = undefined) {
+function assertNearlyEqual(
+  actual: number,
+  expected: number,
```
Comment (statementreply) [inline test/utils.ts]: `expected: number | number[]`

File: test/utils.ts [hunk 5]
```diff
@@ -118,21 +142,34 @@ function nearlyEqual(actual, expected, options = {}) {
...
+  options: {
+    relTol?: number;
+    absTol?: number;
+    alts?: number[];
```
Comment (statementreply) [inline test/utils.ts]: `alts?: number[] | number[][]`

File: test/utils.ts (hunk 1 above)
Comment (antfu) [reply test/utils.ts]: Fixed. Thanks 👍

**Comment by antfu** [comment]
@LingDong- Should I make this into a branch of the main repo and make the PR then? Maybe it's better to try out and make changes to this migration.

**Comment by LingDong-** [comment]
Actually, I'm already convinced. Let's go for it!
Feel free to merge it, turn it into a branch, etc. whenever you feel it's ready :)

**Comment by antfu** [comment]
It's ready. Let's merge then. 🎉 If there is anything wrong, feel free to revert it.

**Pull request merged**